File: VB/E2311/Helper/MyNavBarSuperTipHelper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraNavBar
Imports DevExpress.Utils
Imports System.Drawing.Design

Namespace E2311
	Public Class MyNavBarSuperTipHelper
		Inherits Component

		Public Sub New()
			_Items = New MyNavBarSuperToolTipItemsCollection()
			AddHandler ToolTipController.DefaultController.GetActiveObjectInfo, AddressOf DefaultController_GetActiveObjectInfo
		End Sub

		Private Sub DefaultController_GetActiveObjectInfo(ByVal sender As Object, ByVal e As ToolTipControllerGetActiveObjectInfoEventArgs)
			Dim navBar As NavBarControl = TryCast(e.SelectedControl, NavBarControl)
			If navBar IsNot Nothing Then
				Dim hi As NavBarHitInfo = navBar.CalcHitInfo(e.ControlMousePosition)

				If hi.InLink Then
					Dim superTip As New SuperToolTip()
					Dim args As New SuperToolTipSetupArgs()
					args.Title.Text = hi.Link.Caption
					superTip.Setup(args)
					e.Info.SuperTip = superTip
				Else
					If hi.InGroup Then
						Dim superTip As SuperToolTip = GetSuperTipByGroup(hi.Group)
							If e.Info Is Nothing Then
								e.Info = New ToolTipControlInfo(e.SelectedControl, "")
							End If
							e.Info.ToolTipType = ToolTipType.SuperTip
							e.Info.SuperTip = superTip

					End If
				End If
			End If
		End Sub

		Public Function GetSuperTipByGroup(ByVal group As NavBarGroup) As SuperToolTip
			If group Is Nothing Then
				Return Nothing
			End If
			For Each item As MyNavBarSuperToolTipHelperItem In Items
				If item.Group Is group Then
					Return item.SuperTip
				End If
			Next item
			Return Nothing
		End Function

		Private _Items As MyNavBarSuperToolTipItemsCollection

		<DesignerSerializationVisibility(DesignerSerializationVisibility.Content)> _
		Public Property Items() As MyNavBarSuperToolTipItemsCollection
			Get
				Return _Items
			End Get
			Set(ByVal value As MyNavBarSuperToolTipItemsCollection)
				_Items = value
			End Set
		End Property

		Protected Overrides Overloads Sub Dispose(ByVal disposing As Boolean)
			RemoveHandler ToolTipController.DefaultController.GetActiveObjectInfo, AddressOf DefaultController_GetActiveObjectInfo
			MyBase.Dispose(disposing)
		End Sub
	End Class
End Namespace

File: CS/E2311/Helper/MyNavBarSuperTipHelper.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraNavBar;
using DevExpress.Utils;
using System.Drawing.Design;

namespace E2311
{
    public class MyNavBarSuperTipHelper : Component
    {

        public MyNavBarSuperTipHelper()
        {
            _Items = new MyNavBarSuperToolTipItemsCollection();
            ToolTipController.DefaultController.GetActiveObjectInfo += DefaultController_GetActiveObjectInfo;
        }

        void DefaultController_GetActiveObjectInfo(object sender, ToolTipControllerGetActiveObjectInfoEventArgs e)
        {
            NavBarControl navBar = e.SelectedControl as NavBarControl;
            if (navBar != null)
            {
                NavBarHitInfo hi = navBar.CalcHitInfo(e.ControlMousePosition);

                if (hi.InLink)
                {
                    SuperToolTip superTip = new SuperToolTip();
                    SuperToolTipSetupArgs args = new SuperToolTipSetupArgs();
                    args.Title.Text = hi.Link.Caption;
                    superTip.Setup(args);
                    e.Info.SuperTip = superTip;
                }
                else
                    if (hi.InGroup)
                    {
                        SuperToolTip superTip = GetSuperTipByGroup(hi.Group);
                        {
                            if (e.Info == null)
                                e.Info = new ToolTipControlInfo(e.SelectedControl, "");
                            e.Info.ToolTipType = ToolTipType.SuperTip;
                            e.Info.SuperTip = superTip;
                        }

                    }
            }
        }

        public SuperToolTip GetSuperTipByGroup(NavBarGroup group)
        {
            if (group == null) return null;
            foreach (MyNavBarSuperToolTipHelperItem item in Items)
            {
                if (item.Group == group) return item.SuperTip;
            }
            return null;
        }

        private MyNavBarSuperToolTipItemsCollection _Items;

        [DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public MyNavBarSuperToolTipItemsCollection Items
        {
            get { return _Items; }
            set { _Items = value; }
        }

        protected override void Dispose(bool disposing)
        {
            ToolTipController.DefaultController.GetActiveObjectInfo -= DefaultController_GetActiveObjectInfo;
            base.Dispose(disposing);
        }
    }
}
